velox/functions/prestosql/aggregates/MinMaxByAggregates.cpp

```diff
@@ -628,6 +640,458 @@ class MinByAggregate : public MinMaxByAggregate<T, U> {
+    const auto* comparisonArray = result.comparisonArray;
+    const auto* comparisons = result.comparisons;
+
+    const auto numValues = valueArray->sizeAt(decodedIndex);
+    const auto valueOffset = valueArray->offsetAt(decodedIndex);
+    const auto comparisonOffset = comparisonArray->offsetAt(decodedIndex);
+    for (auto j = 0; j < numValues; ++j) {
+      const auto comparison = comparisons->valueAt(comparisonOffset + j);
+      const auto value = optionalValue(*values, valueOffset + j);
+      accumulator->compareAndAdd(comparison, value, comparator_);
+    }
+  }
+
+  IntermediateResult decodeIntermediateResults(
+      const VectorPtr& arg,
+      const SelectivityVector& rows) {
+    decodedIntermediates_.decode(*arg, rows);
+
+    auto baseRowVector =
+        dynamic_cast<const RowVector*>(decodedIntermediates_.base());
+
+    decodedN_.decode(*baseRowVector->childAt(0), rows);
+    decodedValue_.decode(*baseRowVector->childAt(1), rows);
+    decodedComparison_.decode(*baseRowVector->childAt(2), rows);
+
+    IntermediateResult result;
+    result.valueArray = decodedValue_.base()->template as<ArrayVector>();
+    result.comparisonArray =
+        decodedComparison_.base()->template as<ArrayVector>();
+
+    result.values = result.valueArray->elements()->template as<FlatVector<V>>();
+    result.comparisons =
+        result.comparisonArray->elements()->template as<FlatVector<C>>();
+
+    return result;
+  }
+
+  /// Return total number of values in all accumulators of specified 'groups'.
+  /// Set null flags in 'result'.
+  vector_size_t countValuesAndSetResultNulls(
+      char** groups,
+      int32_t numGroups,
+      VectorPtr& result) {
+    vector_size_t numValues = 0;
+
+    uint64_t* rawNulls = getRawNulls(result.get());
+
+    for (auto i = 0; i < numGroups; ++i) {
+      auto* group = groups[i];
+      auto* accumulator = value(group);
+
+      if (isNull(group)) {
+        result->setNull(i, true);
+      } else {
+        clearNull(rawNulls, i);
+        numValues += accumulator->topPairs.size();
+      }
+    }
+
+    return numValues;
+  }
+
+  int64_t
+  validateN(DecodedVector& decodedN, vector_size_t row, int64_t currentN) {
+    VELOX_USER_CHECK(
+        !decodedN.isNullAt(row),
+        "third argument of max_by/min_by must be a positive integer");
+    const auto n = decodedN.valueAt<int64_t>(row);
+    VELOX_USER_CHECK_GT(
+        n, 0, "third argument of max_by/min_by must be a positive integer");
+
+    if (currentN) {
+      VELOX_USER_CHECK_EQ(
+          n,
+          currentN,
+          "third argument of max_by/min_by must be a constant for all rows in a group");
+    }
+    return n;
+  }
+
+  int64_t extractN(
+      const VectorPtr& arg,
+      const SelectivityVector& rows,
+      int64_t currentN) {
+    decodedN_.decode(*arg, rows);
+    if (decodedN_.isConstantMapping()) {
+      return validateN(decodedN_, rows.begin(), currentN);
+    }
+
+    VELOX_USER_CHECK(
```

Comment:
Why we need check the first row here? Can we just rely on "rows.applyToSelected" at l1070 to handle the check for each row? Is it an optimization for fast abort if the first element fails or? Thanks!